STUDENT PERFORMANCE INDICATOR

--->LIFE CYCLE of Machine learning Project

a. Understanding the Problem Statement

b. Data Collection

c. Data Checks to perform

d. Exploratory data analysis

e. Data Pre-Processing

f. Model Training

g. Choose best model

h. deployement

1) PROBLEM STATEMENT-->
This project understands how the student's performance (test scores) is affected by other variables such as Gender, Ethnicity, Parental level of education, Lunch and Test preparation course.

2) DATA COLLECTION-->
Dataset Source - https://www.kaggle.com/datasets/spscientist/students-performance-in-exams?datasetId=74977
The data consists of 8 column and 1000 rows.

2.1-->IMPORTING MODULES

In [2]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')

Import CSV data as Pandas DataFrame and show 5 records

In [3]:
df = pd.read_csv('data/stud.csv')
df.head(50)

gender race_ethnicity parental_level_of_education         lunch  \
0   female        group B           bachelor's degree      standard   
1   female        group C                some college      standard   
2   female        group B             master's degree      standard   
3     male        group A          associate's degree  free/reduced   
4     male        group C                some college      standard   
5   female        group B          associate's degree      standard   
6   female        group B                some college      standard   
7     male        group B                some college  free/reduced   
8     male        group D                 high school  free/reduced   
9   female        group B                 high school  free/reduced   
10    male        group C          associate's degree      standard   
11    male        group D          associate's degree      standard   
12  female        group B                 high school      standard   
13    male        group A                some college      standard   
14  female        group A             master's degree      standard   
15  female        group C            some high school      standard   
16    male        group C                 high school      standard   
17  female        group B            some high school  free/reduced   
18    male        group C             master's degree  free/reduced   
19  female        group C          associate's degree  free/reduced   
20    male        group D                 high school      standard   
21  female        group B                some college  free/reduced   
22    male        group D                some college      standard   
23  female        group C            some high school      standard   
24    male        group D           bachelor's degree  free/reduced   
25    male        group A             master's degree  free/reduced   
26    male        group B                some college      standard   
27  female        group C           bachelor's degree      standard   
28    male        group C                 high school      standard   
29  female        group D             master's degree      standard   
30  female        group D                some college      standard   
31  female        group B                some college      standard   
32  female        group E             master's degree  free/reduced   
33    male        group D                some college      standard   
34    male        group E                some college      standard   
35    male        group E          associate's degree      standard   
36  female        group D          associate's degree      standard   
37  female        group D            some high school  free/reduced   
38  female        group D          associate's degree  free/reduced   
39    male        group B          associate's degree  free/reduced   
40    male        group C          associate's degree  free/reduced   
41  female        group C          associate's degree      standard   
42  female        group B          associate's degree      standard   
43    male        group B                some college  free/reduced   
44  female        group E          associate's degree  free/reduced   
45    male        group B          associate's degree      standard   
46  female        group A          associate's degree      standard   
47  female        group C                 high school      standard   
48  female        group D          associate's degree  free/reduced   
49    male        group C                 high school      standard   

   test_preparation_course  math_score  reading_score  writing_score  
0                     none          72             72             74  
1                completed          69             90             88  
2                     none          90             95             93  
3                     none          47             57             44  
4                     none 

 Shape Of Data

In [4]:
df.shape

(1000, 8)

2.2---> DATASET INFO

gender : sex of students -> (Male/female)

race/ethnicity : ethnicity of students -> (Group A, B,C, D,E)

parental level of education : parents' final education ->(bachelor's degree,some college,master's degree,associate's degree,high school)

lunch : having lunch before test (standard or free/reduced)

test preparation course : complete or not complete before test

math score

reading score

writing score

3--> Data Checks to perform

Check Missing values

Check Duplicates

Check data type

Check the number of unique values of each column

Check statistics of data set

Check various categories present in the different categorical column

3.a--> Check Missing values

In [5]:
df.isna().sum()

gender                         0
race_ethnicity                 0
parental_level_of_education    0
lunch                          0
test_preparation_course        0
math_score                     0
reading_score                  0
writing_score                  0
dtype: int64

There are no missing values in the data set

3.b-->Check Duplicates

In [7]:
df.duplicated().sum()

0

There are no duplicate values in the data set

3.c--> Check data types

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race_ethnicity               1000 non-null   object
 2   parental_level_of_education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test_preparation_course      1000 non-null   object
 5   math_score                   1000 non-null   int64 
 6   reading_score                1000 non-null   int64 
 7   writing_score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


3.d--> Checking the number of unique values of each column

In [10]:
df.nunique()

gender                          2
race_ethnicity                  5
parental_level_of_education     6
lunch                           2
test_preparation_course         2
math_score                     81
reading_score                  72
writing_score                  77
dtype: int64

3.e--> Check statistics of data set

In [11]:
df.describe()

math_score  reading_score  writing_score
count  1000.00000    1000.000000    1000.000000
mean     66.08900      69.169000      68.054000
std      15.16308      14.600192      15.195657
min       0.00000      17.000000      10.000000
25%      57.00000      59.000000      57.750000
50%      66.00000      70.000000      69.000000
75%      77.00000      79.000000      79.000000
max     100.00000     100.000000     100.000000

Insights

From above description of numerical data, all means are very close to each other - between 66 and 68.05;

All standard deviations are also close - between 14.6 and 15.19;

While there is a minimum score 0 for math, for writing minimum is  10 and for reading minimum score is higher = 17

3.f-->Exploring Data

In [12]:
df.head(10)

gender race_ethnicity parental_level_of_education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   
5  female        group B          associate's degree      standard   
6  female        group B                some college      standard   
7    male        group B                some college  free/reduced   
8    male        group D                 high school  free/reduced   
9  female        group B                 high school  free/reduced   

  test_preparation_course  math_score  reading_score  writing_score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75  
5                    none          71             83             78  
6               completed          88             95             92  
7                    none          40             43             39  
8               completed          64             64             67  
9                    none          38             60             50

In [14]:
print("Categories in 'gender' variable:     ",end=" " )
print(df['gender'].unique())

print("Categories in 'race_ethnicity' variable:  ",end=" ")
print(df['race_ethnicity'].unique())

print("Categories in'parental level of education' variable:",end=" " )
print(df['parental_level_of_education'].unique())

print("Categories in 'lunch' variable:     ",end=" " )
print(df['lunch'].unique())

print("Categories in 'test preparation course' variable:     ",end=" " )
print(df['test_preparation_course'].unique())


Categories in 'gender' variable:      ['female' 'male']
Categories in 'race_ethnicity' variable:   ['group B' 'group C' 'group A' 'group D' 'group E']
Categories in'parental level of education' variable: ["bachelor's degree" 'some college' "master's degree" "associate's degree"
 'high school' 'some high school']
Categories in 'lunch' variable:      ['standard' 'free/reduced']
Categories in 'test preparation course' variable:      ['none' 'completed']


In [16]:
# define numerical & categorical columns
numeric_features = [feature for feature in df.columns if df[feature].dtype != 'O']
categorical_features = [feature for feature in df.columns if df[feature].dtype == 'O']

# print columns
print('We have {} numerical features : {}'.format(len(numeric_features), numeric_features))
print('\nWe have {} categorical features : {}'.format(len(categorical_features), categorical_features))

We have 3 numerical features : ['math_score', 'reading_score', 'writing_score']

We have 5 categorical features : ['gender', 'race_ethnicity', 'parental_level_of_education', 'lunch', 'test_preparation_course']
